<!-- GrassBots simulation launch file -->

<launch>
	<!-- Transverse Mercator Projection parameters --> 
	<param name="transverse_mercator_a" value="6378137.0" />
	<param name="transverse_mercator_f" value="0.0033528106647474805" /> 
	<param name="transverse_mercator_false_easting" value="500000.0" />
	<param name="transverse_mercator_scale_factor" value="0.9996" />
	<param name="transverse_mercator_origin_latitude" value="0.0" />
	<param name="transverse_mercator_central_meridian" value="9.0" />
	<param name="transverse_mercator_false_northing" value="0.0" />

	<group ns="fmSensors">
	</group>

	<group ns="fmProcessors">
	</group>

	<group ns="fmDecisionMakers">
		<node pkg="sdu_grassbots_2014" name="mission" type="keyboard_mission_node.py" output="screen">
			<param name="keyboard_sub" value="/fmHMI/keyboard" />
			<param name="hmi_pub" value="/fmDecision/hmi" />
			<param name="automode_pub" value="/fmDecision/automode" />
			<param name="deadman_pub" value="/fmCommand/deadman" />
			<param name="cmd_vel_pub" value="/fmCommand/cmd_vel" />
			<param name="max_linear_velocity" value="1.0"/>
			<param name="max_angular_velocity" value="0.4"/>
			<param name="linear_velocity_step" value="0.1"/>
			<param name="angular_velocity_step" value="0.1"/>
		</node>
	</group>

	<group ns="fmExecutors">
		<node pkg="sdu_grassbots_2014" type="waypoint_navigation_node.py" name="wptnav" output="screen">
			<param name="pose_sub" value="/fmKnowledge/pose" />
			<param name="automode_sub" value="/fmDecision/automode" />
			<param name="hmi_sub" value="/fmDecision/hmi" />
			<param name="routept_pub" value="/fmPlan/route_point" />
			<param name="cmd_vel_pub" value="/fmCommand/cmd_vel" />
			<param name="implement_pub" value="/fmCommand/implement" />
			<param name="status_pub" value="/fmInformation/wptnav_status" />
			<param name="pid_pub" value="/fmInformation/wptnav_pid" />
			<param name="status_publish_interval" value="1" />
			<param name="pid_publish_interval" value="1" />
			<param name="print_debug_information" value="false" />

			<param name="max_linear_velocity" value="1.2"/>
			<param name="max_angular_velocity" value="0.70"/>

			<param name="drive_kp" value="1.8"/>
			<param name="drive_ki" value="0.4"/>
			<param name="drive_kd" value="0.0"/>
			<param name="drive_feed_forward" value="0.00"/>
			<param name="drive_max_output" value="0.50"/>

			<param name="turn_kp" value="0.6"/>
			<param name="turn_ki" value="0.4"/>
			<param name="turn_kd" value="0.1"/>
			<param name="turn_feed_forward" value="0.0"/>
			<param name="turn_max_output" value="1.0"/>

			<param name="wpt_default_mode" value="MCTE" />
			<param name="wpt_default_tolerance" value="0.20" />
			<param name="wpt_default_drive_velocity" value="0.9" />
			<param name="wpt_default_turn_velocity" value="0.7" />
			<param name="wpt_default_wait_after_arrival" value="2.0" />
			<param name="wpt_default_implement_command" value="0.0" />

			<param name="target_ahead" value="1.5" />
			<param name="turn_start_at_heading_err" value="60.0" />
			<param name="turn_stop_at_heading_err" value="2.0" />
			<param name="ramp_drive_velocity_at_distance" value="1.0" />
			<param name="ramp_min_drive_velocity" value="0.15" />
			<param name="ramp_turn_velocity_at_angle" value="30.0" />
			<param name="ramp_min_turn_velocity" value="0.15" />
			<param name="stop_navigating_at_distance" value="0.03" />
		</node>		

		<node pkg="sdu_grassbots_2014" name="route_plan_socketd" type="route_plan_socketd_node.py" output="screen" respawn="true" >
			<param name="hmi_pub" value="/fmDecision/hmi" />
			<param name="routept_pub" value="/fmPlan/route_point" />
			<param name="socket_address" value="localhost" />
			<param name="socket_port" value="8080" />
			<param name="socket_timeout" value="30" />
			<param name="socket_password" value="" />
			<param name="debug" value="false" />
		</node>
	</group>

	<group ns="fmControllers">
	</group>

	<group ns="fmLib">
		<node pkg="frobit_sim" name="froboscout_sim" type="frobit_sim_node.py" output="screen" >
			<param name="deadman_sub" value="/fmCommand/deadman" />
			<param name="cmd_vel_sub" value="/fmCommand/cmd_vel" />
			<param name="odom_reset_sub" value="/fmInformation/odom_reset" />
			<param name="pose_pub" value="/fmKnowledge/pose" />
			<param name="max_linear_acceleration" value="1.0"/>
			<param name="max_angular_acceleration" value="1.0"/>
			<param name="wheel_feedback_left_pub" value="/fmInformation/wheel_feedback_left" />
			<param name="wheel_feedback_right_pub" value="/fmInformation/wheel_feedback_right" />
			<param name="publish_wheel_feedback_rate" value="50" />
		</node>

	 	<node pkg="keyboard_interface" type="keyboard_node" name="keyboard">
			<param name="keyboard_pub" value="/fmHMI/keyboard" />
		</node>

		<!--node pkg="plot_velocity" name="plot_vel" type="plot_velocity_node.py" output="screen" >
 			<param name="velocity_sub" value="/fmCommand/cmd_vel" />
 			<param name="update_frequency" value="10" />
			<param name="plot_title" value="cmd_vel" />
			<param name="max_linear_velocity" value="1.2" />
			<param name="max_angular_velocity" value="1.0" />
			<param name="samples" value="100" />
		</node-->

		<!--node pkg="plot_propulsion_feedback_leftright" name="plot_feedback" type="plot_propulsion_feedback_leftright_node.py" output="screen" >
 			<param name="propulsion_feedback_left_sub" value="/fmInformation/wheel_feedback_left" />
 			<param name="propulsion_feedback_right_sub" value="/fmInformation/wheel_feedback_right" />
 			<param name="update_frequency" value="50" />
			<param name="plot_title" value="Wheel feedback" />
			<param name="maximum_velocity" value="1.2" />
			<param name="maximum_thrust" value="511" />
		</node-->

		<!--node pkg="plot_pid" name="wptnav_pid" type="plot_pid_node.py" output="screen" >
 			<param name="pid_sub" value="/fmInformation/wptnav_pid" />
 			<param name="update_rate" value="10" />
			<param name="plot_title" value="Navigation PID status" />
			<param name="samples" value="100" />
			<param name="max_error" value="0.5" />
			<param name="max_output" value="0.5" />
			<param name="max_pid" value="1.1" />
			<param name="plot_feed_forward" value="true" />
		</node-->

		<node pkg="frobit_plot" name="frobit_plot" type="frobit_plot_node.py" output="screen" >
			<param name="pose_sub" value="/fmKnowledge/pose" />
			<param name="gnss_sub" value="/fmInformation/gpgga_tranmerc" />
			<param name="odom_sub" value="/fmKnowledge/odom" />
			<param name="wptnav_status_sub" value="/fmInformation/wptnav_status" />
			<param name="plot_pose_track" value="true" />
			<param name="plot_gnss_track" value="false" />
			<param name="plot_odometry_track" value="false" />
			<param name="plot_pose_yaw" value="false" />
			<param name="map_trackpoint_threshold" value="0.05" />
			<param name="map_max_trackpoints" value="1000" />
			<param name="map_minimum_size" value="6.0" />
			<param name="map_easting_offset" value="0.0" />
			<param name="map_northing_offset" value="0.0" />
			<param name="map_update_frequency" value="10.0" />
			<param name="map_title" value="GrassBots Track" />
			<param name="map_window_size" value="5.0" />
			<param name="avatar_extension_front" value="1.7" />
			<param name="avatar_extension_rear" value="0.9" />
			<param name="avatar_extension_lateral" value="0.9" />
		</node>
	</group>
</launch>

